File: themes/reverse-string/go/main_test.go
package main

import "testing"

func TestReverseGraphemes(t *testing.T) {
	tests := []struct {
		name     string
		input    string
		expected string
	}{
		{"ASCII", "asdf", "fdsa"},
		{"combining characters", "as⃝df̅", "f̅ds⃝a"},
		{"Japanese", "ようこそ,世界に!", "!に界世,そこうよ"},
		{"ZWJ emoji", "👨‍👩‍👧‍👦", "👨‍👩‍👧‍👦"},
		{"empty string", "", ""},
		{"single character", "a", "a"},
		{"emoji sequence", "🍎🍊🍋", "🍋🍊🍎"},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			got := reverseGraphemes(tt.input)
			if got != tt.expected {
				t.Errorf("reverseGraphemes(%q) = %q, want %q", tt.input, got, tt.expected)
			}
		})
	}
}

File: themes/reverse-string/typescript/tests/main.test.ts
import { describe, expect, test } from "bun:test";
import { reverseGraphemes } from "../src/main";

describe("reverseGraphemes", () => {
  test("ASCII string", () => {
    expect(reverseGraphemes("asdf")).toBe("fdsa");
  });

  test("combining characters", () => {
    // s with combining enclosing circle, f with combining overline
    expect(reverseGraphemes("as⃝df̅")).toBe("f̅ds⃝a");
  });

  test("Japanese", () => {
    expect(reverseGraphemes("ようこそ,世界に!")).toBe("!に界世,そこうよ");
  });

  test("ZWJ emoji", () => {
    // Family emoji (ZWJ sequence) should stay as single cluster
    expect(reverseGraphemes("👨‍👩‍👧‍👦")).toBe("👨‍👩‍👧‍👦");
  });

  test("empty string", () => {
    expect(reverseGraphemes("")).toBe("");
  });

  test("single character", () => {
    expect(reverseGraphemes("a")).toBe("a");
  });

  test("emoji sequence", () => {
    expect(reverseGraphemes("🍎🍊🍋")).toBe("🍋🍊🍎");
  });
});
